docs/rfc/rfc-017-abci++-vote-extension-propag.md

```diff
@@ -0,0 +1,438 @@
+    would still be correct, as all the information a node needs to catch up, and to start proposing when
+    it considers it is caught up, can now be recovered from past blocks.
+
+    This solution has two main drawbacks.
+
+    - It changes the blockchain format, in particular, the way light blocks are verified; thus breaking
+      backwards compatiblity in features such as light clients and IBC (the latter relying on the former).
+    - The extra information (i.e., the vote extensions) that is now kept in the blockchain is not really
+      needed *at every height* for a late node to catch up. This information is only needed to be able
+      to *propose* at the height the validator considers itself as caught-up. If a validator is indeed
+      late for height *h*, it is useless (although correct) for it to call `PrepareProposal`, or
+      `ExtendVote`, since the block is already decided.
+
+- **2.** *Skip* propose *step in Tendermint algorithm*.
+
+    This solution consists in modifying the Tendermint algorithm to skip the *send proposal* step in
+    heights where the node does not have the required vote extensions to populate the call to
+    `PrepareProposal`. The assumption is that the validator is late and, therefore, up-to-date
+    validators have already proposed (and decided) for that height.
+    A small variation of this solution is, rather than skipping the *send proposal* step, the validator
+    sends a special *empty* or *bottom* proposal to signal other nodes that it is not ready to propose
+    at (any round of) the current height.
+
+    The appeal of this solution is its simplicity. A possible implementation does not need to extend
+    the data structures, or change the current catch-up mechanisms implemented in the blocksync or
+    in the consensus reactor. When we lack the needed information (vote extensions), we simply rely
+    on another correct validator to propose a valid block in other rounds of the current height.
+
+    However, this solution can be attacked by a byzantine node in the network in the following way.
+    Let us consider the following scenario:
+
+    - all validators send out precommit messages, with vote extensions, for height *h*, round *r*,
+      roughly at the same time,
+    - all validators then wait until they gather enough precommit messages for height *h*, round *r*,
+      in order to decide in height *h*,
+    - all those precommit messages (and any further message for height *h*) get delayed indefinitely,
+    - an intermediate (malicious) node *m* manages to gather more than *2n/3* of the precommit messages
+      for height *h*, round *r*, as well as the proposed block the precommit messages refer to,
+    - node *m* uses the precommit messages to build a commit (not extended commit), which it then sends
+      to *all* validators in order to convince them they are late,
+    - all validators receive the *catch-up* message from *m*, decide on height *h*, and proceed to
```

Comment:
Can you clarify what this 'catch-up' message is? Does it currently cause Tendermint to behave differently? Perhaps this is just a bit of ignorance on my part.

Reply:
I left it vague on purpose, as we are explaining the attack in abstract terms. The catchup message could just be "hey I am already in height *h+delta*, here's the decided block for *h*"
One way or another, this kind of message has to exist.
Let me know if you have suggestions on how to render this text clearer

Reply:
> There is no consensus message that carries a commit.
That is true but remember I'm trying to explain it here in abstract terms (so if a message doesn't exist, we could just assume it does for the sake of the discussion).

Anyway, after gathering your feedback, I think the solution here is to provide some definition of what I understand as catch-up message early on in the document. I'll do that